buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -110,239 +205,267 @@ int usage() {
+            byte val2 = value(id ^ 1);
+            byte val = val1 < val2 ? val1 : val2;
+            setValue(parentId, val);
+            id = parentId;
+        }
+    }
 
-                curIdx = nextIdx ^ 1;
-                val = memoryMap[curIdx];
-                continue;
+    /**
+     * Update method used by free
+     * This needs to handle the special case when both children are completely free
+     * in which case parent be directly allocated on request of size = child-size * 2
+     *
+     * @param id id
+     */
+    private void updateParentsFree(int id) {
+        int logChild = depth(id) + 1;
+        while (id > 1) {
+            int parentId = id >>> 1;
+            byte val1 = value(id);
+            byte val2 = value(id ^ 1);
+            logChild -= 1; // in first iteration equals log, subsequently reduce 1 from logChild as we traverse up
+
+            if (val1 == logChild && val2 == logChild) {
+                setValue(parentId, (byte) (logChild - 1));
+            } else {
+                byte val = val1 < val2 ? val1 : val2;
+                setValue(parentId, val);
             }
 
-            // state == ST_UNUSED
-            return allocateRunSimple(normCapacity, curIdx, val);
+            id = parentId;
         }
     }
 
-    private long allocateRunSimple(int normCapacity, int curIdx, int val) {
-        int runLength = runLength(val);
-        if (normCapacity > runLength) {
+    /**
+     * Algorithm to allocate an index in memoryMap when we query for a free node
+     * at depth d
+     *
+     * @param d depth
+     * @return index in memoryMap
+     */
+    private int allocateNode(int d) {
+        int id = 1;
+        int initial = - (1 << d); // has last d bits = 0 and rest all = 1
+        byte val = value(id);
+        if (val > d) { // unusable
             return -1;
         }
-
-        for (;;) {
-            if (normCapacity == runLength) {
-                // Found the run that fits.
-                // Note that capacity has been normalized already, so we don't need to deal with
-                // the values that are not power of 2.
-                memoryMap[curIdx] = val & ~3 | ST_ALLOCATED;
-                freeBytes -= runLength;
-                return curIdx;
+        while (val < d || (id & initial) == 0) { // id & initial == 1 << d for all ids at depth d, for < d it is 0
+            id <<= 1;
+            val = value(id);
+            if (val > d) {
+                id ^= 1;
+                val = value(id);
             }
-
-            int nextIdx = curIdx << 1 ^ nextRandom();
-            int unusedIdx = nextIdx ^ 1;
-
-            memoryMap[curIdx] = val & ~3 | ST_BRANCH;
-            //noinspection PointlessBitwiseExpression
-            memoryMap[unusedIdx] = memoryMap[unusedIdx] & ~3 | ST_UNUSED;
-
-            runLength >>>= 1;
-            curIdx = nextIdx;
-            val = memoryMap[curIdx];
         }
+        byte value = value(id);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "value". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)